pkg/geo/geomfn/distance.go

```diff
@@ -814,3 +830,149 @@ func verifyDensifyFrac(f float64) error {
 	}
 	return nil
 }
+
+// PointPolygonRelationType defines a relationship type between
+// a (multi)point and a (multi)polygon.
+type PointPolygonRelationType bool
+
+const (
+	// PointPolygonIntersects is the relationship where a (multi)point
+	// intersects a (multi)polygon.
+	PointPolygonIntersects PointPolygonRelationType = true
+	// PointPolygonWithin is the relationship where a (multi)point
+	// is within a (multi)polygon.
+	PointPolygonWithin PointPolygonRelationType = false
+)
+
+// PointKindRelatesToPolygonKind returns whether a (multi)point and
+// a (multi)polygon have the given relationship.
+func PointKindRelatesToPolygonKind(
+	pointKind geo.Geometry, polygonKind geo.Geometry, relationType PointPolygonRelationType,
+) (bool, error) {
+	pointKindBaseT, err := pointKind.AsGeomT()
+	if err != nil {
+		return false, err
+	}
+	polygonKindBaseT, err := polygonKind.AsGeomT()
+	if err != nil {
+		return false, err
+	}
+	pointKindIterator := geo.NewGeomTIterator(pointKindBaseT, geo.EmptyBehaviorOmit)
+	polygonKindIterator := geo.NewGeomTIterator(polygonKindBaseT, geo.EmptyBehaviorOmit)
+
+	// Check whether each point intersects with at least one polygon.
+	// For Intersects, at least one point must intersect with at least one polygon.
+	// For Within, every point must intersect with at least one polygon.
+	ret := false
+pointOuterLoop:
+	for {
+		point, hasPoint, err := pointKindIterator.Next()
+		if err != nil {
+			return false, err
+		}
+		if !hasPoint {
+			break
+		}
+		pointGeometry, err := geo.MakeGeometryFromGeomT(point)
+		if err != nil {
+			return false, err
+		}
+		// Reset the polygon iterator on each iteration of the point iterator.
+		polygonKindIterator.Reset()
+		for {
+			polygon, hasPolygon, err := polygonKindIterator.Next()
+			if err != nil {
+				return false, err
+			}
+			if !hasPolygon {
+				break
+			}
+			polygonGeometry, err := geo.MakeGeometryFromGeomT(polygon)
+			if err != nil {
+				return false, err
+			}
+			intersects, err := PointIntersectsPolygon(pointGeometry, polygonGeometry)
```

Comment:
I don't think within / contains is a strict intersection -- it *cannot* be on any of the edges.

see https://mentin.medium.com/which-predicate-cb608b470471.

Reply:
(i would say if you can, do within/contains in a separate PR. you can keep the tests -- they should still pass)

Reply:
yes, but as commented earlier, please add this in a separate PR!

Reply:
http://lin-ear-th-inking.blogspot.com/2007/06/subtleties-of-ogc-covers-spatial.html

i need to read this every time, lol

Reply:
Added the PIP optimization to st_covers in #62582